libs/wire-api/src/Wire/API/UserEvent.hs

```diff
@@ -0,0 +1,451 @@
+                              <$> eirId .= field "id" schema
+                              <*> eirEmail .= maybe_ (optField "email" schema)
+                              <*> eirPhone .= maybe_ (optField "phone" schema)
+                          )
+                      )
+                  )
+              )
+          EventTypeUserSuspended -> tag _UserEvent (tag _UserSuspended (field "id" schema))
+          EventTypeUserResumed -> tag _UserEvent (tag _UserResumed (field "id" schema))
+          EventTypeUserDeleted ->
+            tag
+              _UserEvent
+              ( tag
+                  _UserDeleted
+                  ( field "qualified_id" schema
+                      <* qUnqualified .= field "id" schema
+                  )
+              )
+          EventTypeUserLegalholdEnabled ->
+            tag
+              _UserEvent
+              ( tag _UserLegalHoldEnabled (field "id" schema)
+              )
+          EventTypeUserLegalholdDisabled ->
+            tag
+              _UserEvent
+              ( tag _UserLegalHoldDisabled (field "id" schema)
+              )
+          EventTypeUserLegalholdRequested ->
+            tag
+              _UserEvent
+              ( tag
+                  _LegalHoldClientRequested
+                  ( LegalHoldClientRequestedData
+                      <$> lhcTargetUser .= field "id" schema
+                      <*> lhcLastPrekey .= field "last_prekey" schema
+                      <*> lhcClientId .= field "client" (idObjectSchema schema)
+                  )
+              )
+          EventTypePropertiesSet ->
+            tag
+              _PropertyEvent
+              ( tag
+                  _PropertySet
+                  ( (,)
+                      <$> fst .= field "key" genericToSchema
+                      <*> snd .= field "value" jsonValue
+                  )
+              )
+          EventTypePropertiesDeleted ->
+            tag
+              _PropertyEvent
+              ( tag
+                  _PropertyDeleted
+                  (field "key" genericToSchema)
+              )
+          EventTypePropertiesCleared ->
+            tag
+              _PropertyEvent
+              ( tag
+                  _PropertiesCleared
+                  (pure ())
+              )
+          EventTypeClientAdded ->
+            tag
+              _ClientEvent
+              ( tag
+                  _ClientAdded
+                  (field "client" schema)
+              )
+          EventTypeClientRemoved ->
+            tag
+              _ClientEvent
+              ( tag
+                  _ClientRemoved
+                  (field "client" (idObjectSchema schema))
+              )
+          EventTypeConnection ->
+            tag
+              _ConnectionEvent
+              ( ConnectionUpdated
+                  <$> ucConn .= field "connection" schema
+                  <*> ucName .= maybe_ (optField "user" (object "UserName" (field "name" schema)))
+              )
```

Comment:
A lot of nesting, but I guess it's fine :sweat_smile:

Reply:
It follows the structure of the type. I don't really see anything wrong with this.